### DISCUSSION TITLE
Handling List Type with Client Side Symmetric Encryption

### DISCUSSION THREAD

**Discussion by kakedis** [sqlalchemy/sqlalchemy#10000] [Usage Questions]
I've implemented Client Side Symmetric Encryption with great success based on the [TypeDecorator example](https://github.com/sqlalchemy/sqlalchemy/wiki/SymmetricEncryptionClientSide#example-two---use-typedecorator). However, I'm running into some issues when trying to encrypt list type (python). This is how I'm performing the encryption (based on the example linked above);
```
def encrypt(self, data):
        cipher = AES.new(self.key, AES.MODE_EAX, nonce=self.nonce)
        data = data + (" " * (16 - (len(data) % 16)))
        return cipher.encrypt(data.encode("utf-8")).hex()
```
Whenever the value is a list I get the following exception:
`TypeError: can only concatenate list (not "str") to list
`

The error is occurring in the padding part of the encryption;
`data = data + (" " * (16 - (len(data) % 16)))
`

I have a couple of questions on this;

1. Would I be able to fix this issue by doing the encoding before adding the padding or is there a better way about this?
2. I already have a bunch of data encrypted, so I'm worried that if I mess with the padding or anything else it would break the encrypt/decrypt for existing data. I think I can probably handle both sets of encryption (old handling just strings and new handling list), but I'm worried that this won't be the most optimized approached.

**Comment by CaselIT** [comment]
Hi,

I'm not sure why the data would ever be a list. In general you cannot save arrays in a database column, so assigning one to a column would be a strange use case.

Can you clarify your use case?

**Reply by kakedis** [reply]
Thanks for the reply @CaselIT I'm not sure what the original thought was behind saving a the data in the db as list, but I assume its due to how it's getting consumed. The list contains a bunch of long urls that are strings within the list. So it looks something like this;
`['...url...', '...url...', '...url...', '...url...']` 
and saved in a varchar field. It gets consumed by a third-party integration API and the value just gets passed on like this. I can look into changing the value being saved to the db, but I'm worried about legacy data and it touches a few places, hence the reason I'm trying to solve it from the encryption side, so that I can continue using the list as value in the varchar field. 

I understand its not an ideal situation. Any help or suggestions will be appreciated!

**Reply by CaselIT** [reply]
You can't save a list of strings in a varchar field, that's what I'm saying above. So you will have to decide how to serialize the list to strings to a single string that can be saved in the varchar fiel. Once you have the string it's not any different than what you're doing now

**Comment by zzzeek** [comment]
issue number 10000!

**Reply by kakedis** [reply]
Yay, glad I was able to contribute!
